framework bug fixed; with added benchmark for error checking

lijun99 opened the pull request:
This request includes all necessary changes in order for altar to produce the correct results. 
The added benchmark serves also as an error checking.

Commit:
framework bug fixed; with added benchmark for error checking

aivazis commented on altar/utils/__init__.py:
```diff
@@ -0,0 +1,22 @@
+import altar
+
+# import utitlities for gsl matrix/vector to ndarray conversion
+from .Ndarray import *
```
this is very heavily discouraged; please import explicitly the symbols that you intend to publish

lijun99 replied:
This can be easily fixed, once we decide the best place to put these utilities.

aivazis commented on altar/utils/__init__.py:
```diff
@@ -0,0 +1,22 @@
+from .Ndarray import *
+
+# import mean/sd utitilies for matrix/vector
+from .Stats import *
```
this is very heavily discouraged; please import explicitly the symbols that you intend to publish

aivazis commented on altar/bayesian/Metropolis.py:
```diff
@@ -29,7 +29,7 @@ class Metropolis(altar.component, family="altar.samplers.metropolis", implements
 
     # user configurable state
-    scaling = altar.properties.float(default=.1)
+    scaling = altar.properties.float(default=.3)
```
i'm not sure I'm understand the value of changing the default here; the user has complete control over this. why is 0.3 any better than 0.1?

lijun99 replied:
In version 1.1, the default value for scaling is for kc*kc=0.1. In 2.0, scaling is redefined as kc, so we should use sqrt(0.1) ~ 0.3 (or 0.316). Too small scaling in the beginning step will limit the search range of parameter space, and could lead to local optima. This has been observed in some tests, especially with small number of chains and/or small number of chain steps. I believe Sarah et al has studied this factor well, to set it as 0.1 [or sqrt(0.1) by new definition]. 

Of course, this is a user-configurable value as long as they know what value to use. Otherwise, we should offer an optimal value as default.

aivazis commented on altar/bayesian/Metropolis.py:
```diff
@@ -50,7 +50,7 @@ def initialize(self, application):
         rng = application.rng.rng
         # set up the distribution for building the sample multiplicities
-        self.uniform = altar.pdf.uniform(support=(0,1), rng=rng)
+        self.uniform = altar.pdf.uniform_pos(rng=rng)
```
what problem is being fixed here?

lijun99 replied:
This is to use a positive uniform distribution (0,1) instead of regular uniform generator [0,1). Since we use its log value, a zero random number will break the program. A zero random number is really rare, but indeed happened once or twice to me.

aivazis commented on altar/norms/L2.py:
```diff
@@ -49,7 +49,7 @@ def withCovariance(self, v, sigma_inv):
             sigma_inv.lowerTriangular, sigma_inv.opNoTrans, sigma_inv.nonUnitDiagonal,
             sigma_inv, v)
         # compute the dot product and return it
-        return altar.blas.dnrm2(v)
```
are you sure you are not dropping a factor of 2 in the final answer here?

lijun99 replied:
The factor 2  ( 1/2 ||x||^2 ) in your linear model is given in Model/dataLikelihood method. Of course, we could move it here (L2.py).

aivazis commented on altar/utils/Ndarray.py:
```diff
@@ -0,0 +1,53 @@
+# -*- python -*-
```
i'm not sure altar is the correct place for a GSL to numpy adapter; i would like to avoid opening this can of worms. would you consider contributing this to the pyre GSL package instead? it might need a bit of clean up, but i think it's a better place for it

lijun99 replied:
I agree that pyre/gsl is a better place for it.

aivazis commented on altar/utils/Stats.py:
```diff
@@ -0,0 +1,87 @@
+# -*- python -*-
```
this filename violates the filename conventions. since there is no class Stats in it, it should be all lowercase

aivazis replied:
also, this is best done as part of the GSL bindings. both vector and matrix might benefit from such support

aivazis replied:
i'm not sure i like the type universality here: do one thing for vectors, do something else, similar but not the same, for matrices; this can benefit from more thinking

lijun99 replied:
I will add these statistical utilities to pyre/gsl/matrix class.

lijun99 commented:
Hi Michael,

I have addressed your comments on github. Please check.

Now the issue is where to put the utils,
1) a numpy array hook.
I use it mainly for hooking up with the h5py package. But as I believe
while numpy is widely used, it is useful to provide a wrapper or conversion
tools.
2) Stats,
I will move the contents as methods of gsl/matrix in pyre,

If you agree, I will do the changes 1) 2) in pyre at first and submit a PR.

Best,
Lijun



On Sun, Jan 6, 2019 at 9:33 PM Michael Aivazis <notifications@github.com>
wrote:

> *@aivazis* requested changes on this pull request.
> ------------------------------
>
> In altar/utils/__init__.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244434397>:
>
> > +# -*- python -*-
> +# -*- coding: utf-8 -*-
> +#
> +# Lijun Zhu
> +#
> +# (c) 2013-2018 parasim inc
> +# (c) 2010-2018 california institute of technology
> +# all rights reserved
> +#
> +
> +
> +# the package
> +import altar
> +
> +# import utitlities for gsl matrix/vector to ndarray conversion
> +from .Ndarray import *
>
> this is very heavily discouraged; please import explicitly the symbols
> that you intend to publish
> ------------------------------
>
> In altar/utils/__init__.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244434426>:
>
> > +# Lijun Zhu
> +#
> +# (c) 2013-2018 parasim inc
> +# (c) 2010-2018 california institute of technology
> +# all rights reserved
> +#
> +
> +
> +# the package
> +import altar
> +
> +# import utitlities for gsl matrix/vector to ndarray conversion
> +from .Ndarray import *
> +
> +# import mean/sd utitilies for matrix/vector
> +from .Stats import *
>
> this is very heavily discouraged; please import explicitly the symbols
> that you intend to publish
> ------------------------------
>
> In altar/bayesian/Metropolis.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244434720>:
>
> > @@ -29,7 +29,7 @@ class Metropolis(altar.component, family="altar.samplers.metropolis", implements
>
>
>      # user configurable state
> -    scaling = altar.properties.float(default=.1)
> +    scaling = altar.properties.float(default=.3)
>
> i'm not sure I'm understand the value of changing the default here; the
> user has complete control over this. why is 0.3 any better than 0.1?
> ------------------------------
>
> In altar/bayesian/Metropolis.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244434753>:
>
> > @@ -50,7 +50,7 @@ def initialize(self, application):
>          # get the capsule of the random number generator
>          rng = application.rng.rng
>          # set up the distribution for building the sample multiplicities
> -        self.uniform = altar.pdf.uniform(support=(0,1), rng=rng)
> +        self.uniform = altar.pdf.uniform_pos(rng=rng)
>
> what problem is being fixed here?
> ------------------------------
>
> In altar/norms/L2.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244434817>:
>
> > @@ -49,7 +49,7 @@ def withCovariance(self, v, sigma_inv):
>              sigma_inv.lowerTriangular, sigma_inv.opNoTrans, sigma_inv.nonUnitDiagonal,
>              sigma_inv, v)
>          # compute the dot product and return it
> -        return altar.blas.dnrm2(v)
>
> are you sure you are not dropping a factor of 2 in the final answer here?
> ------------------------------
>
> In altar/utils/Ndarray.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244434970>:
>
> > @@ -0,0 +1,53 @@
> +# -*- python -*-
>
> i'm not sure altar is the correct place for a GSL to numpy adapter; i
> would like to avoid opening this can of worms. would you consider
> contributing this to the pyre GSL package instead? it might need a bit of
> clean up, but i think it's a better place for it
> ------------------------------
>
> In altar/utils/Stats.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244435047>:
>
> > @@ -0,0 +1,87 @@
> +# -*- python -*-
>
> this filename violates the filename conventions. since there is no class
> Stats in it, it should be all lowercase
> ------------------------------
>
> In altar/utils/Stats.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244435151>:
>
> > @@ -0,0 +1,87 @@
> +# -*- python -*-
>
> also, this is best done as part of the GSL bindings. both vector and
> matrix might benefit from such support
> ------------------------------
>
> In altar/utils/Stats.py
> <https://github.com/AlTarFramework/altar/pull/9#discussion_r244435573>:
>
> > @@ -0,0 +1,87 @@
> +# -*- python -*-
>
> i'm not sure i like the type universality here: do one thing for vectors,
> do something else, similar but not the same, for matrices; this can benefit
> from more thinking
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AlTarFramework/altar/pull/9#pullrequestreview-188387800>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ak8ZGajIieDPMA74Hx-VADhQWOBtn7EOks5vAtw0gaJpZM4ZblkZ>
> .
>
